pallets/cash/src/internal/events.rs

```diff
@@ -1,182 +1,416 @@
+                        event.clone(),
+                        reason,
+                    ));
+                }
             }
+            return false; // remove from queue
+        } else {
+            return true; // retain on queue
+        }
+    });
+
+    Ok(())
+}
+
+/// Submit the underlying chain blocks the worker calculates are needed by the chain next.
+pub fn submit_chain_blocks<T: Config>(
+    chain_id: ChainId,
+    blocks: ChainBlocks,
+) -> Result<(), Reason> {
+    if blocks.len() > 0 {
+        // XXX TODO whatever we submit we must store first so we can formulate reorgs
+        //  and worry about pruning history independently / later
+        log!("Submitting chain blocks extrinsic: {:?}", blocks);
+        let signature = validator_sign::<T>(&blocks.encode()[..])?;
+        let call = Call::receive_chain_blocks(blocks, signature);
+        if let Err(err) = SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call.into())
+        {
+            log!("Error while submitting chain blocks: {:?}", err);
+            return Err(Reason::FailedToSubmitExtrinsic);
         }
     }
+
     Ok(())
 }
 
-fn submit_events<T: Config>(events: Vec<(ChainLogId, ChainLogEvent)>) -> Result<(), Reason> {
-    for (event_id, event) in events.into_iter() {
-        log!(
-            "Processing event and sending extrinsic: {} {:?}",
-            event_id.show(),
-            event
-        );
-
-        // XXX why are we signing with eth?
-        //  bc eth is identity key...
-        let signature = <Ethereum as Chain>::sign_message(&event.encode()[..])?;
-        let call = Call::receive_event(event_id, event, signature);
-
-        // TODO: Do we want to short-circuit on an error here?
-        let res = SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call.into())
-            .map_err(|()| Reason::FailedToSubmitExtrinsic);
-
-        if res.is_err() {
-            log!("Error while sending event extrinsic");
+/// Walk backwards through the locally stored blocks, in order to formulate a reorg path.
+pub fn build_blocks_backwards<T: Config>(
+    start_hash: ChainHash,
+    num_blocks: u32,
+) -> Result<Vec<ChainBlock>, Reason> {
+    // XXX double check pathological corner cases when there aren't enough blocks
+    // XXX let mut block = fetch start hash from storage
+    for i in 0..num_blocks {
+        // XXX block = fetch parent_hash of block from storage
+    }
+    Ok(vec![])
+}
+
+/// Try to form a path from the last block to the new true block.
+pub fn formulate_reorg<T: Config>(
+    chain_id: ChainId,
+    last_block: &ChainBlock,
+    true_block: &ChainBlock,
+) -> Result<ChainReorg, Reason> {
+    // XXX TODO whatever we submit we must store first so we can formulate reorgs
+    //  and worry about pruning history independently / later
+    const CHUNK: u32 = 10;
+    let mut reverse_blocks: Vec<ChainBlock> = vec![];
+    let mut drawrof_blocks: Vec<ChainBlock> = vec![];
```

Comment:
What does `drawrof` mean?

Reply:
Its the `forward_blocks` in reverse order (needs to be reversed at the end: `collect_rev`)